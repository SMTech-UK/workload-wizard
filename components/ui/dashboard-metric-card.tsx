import { ReactNode } from "react";

interface DashboardMetricCardProps {
  title: string;
  value: ReactNode;
  subtitle?: string;
  icon?: ReactNode;
  valueClassName?: string;
  children?: ReactNode; // For custom content like Progress bar
}

export function DashboardMetricCard({
  title,
  value,
  subtitle,
  icon,
  valueClassName = "",
  children,
}: DashboardMetricCardProps) {
  return (
    <div className="rounded-lg border p-6 bg-white shadow-sm">
      <div className="flex flex-row items-center justify-between space-y-0 pb-2">
        <span className="text-sm font-medium text-gray-500">{title}</span>
        {icon}
      </div>
      <div className={`text-2xl font-bold ${valueClassName}`}>{value}</div>
      {children}
      {subtitle && <div className="text-xs text-muted-foreground mt-1">{subtitle}</div>}
    </div>
  );
} 