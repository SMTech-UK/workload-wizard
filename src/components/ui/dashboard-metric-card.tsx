import { ReactNode } from "react";

interface DashboardMetricCardProps {
  title: string;
  value: ReactNode;
  subtitle?: string;
  icon?: ReactNode;
  valueClassName?: string;
  children?: ReactNode; // For custom content like Progress bar
}

export function DashboardMetricCard({
  title,
  value,
  subtitle,
  icon,
  valueClassName = "",
  children,
}: DashboardMetricCardProps) {
  return (
    <div className="rounded-lg border p-6 bg-white dark:bg-zinc-900 shadow-sm">
      <div className="flex flex-row items-center justify-between space-y-0 pb-2">
        <span className="text-sm font-medium text-gray-900 dark:text-white">{title}</span>
        {icon}
      </div>
      <div className={`text-2xl font-bold ${valueClassName}`}>{value}</div>
      {children}
      {subtitle && <div className="text-xs text-gray-600 dark:text-gray-300 mt-1">{subtitle}</div>}
    </div>
  );
} 