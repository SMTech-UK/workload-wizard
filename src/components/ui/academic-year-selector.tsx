"use client"

import React from "react";
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "@/components/ui/select";
import { Badge } from "@/components/ui/badge";
import { useAcademicYear } from "@/hooks/useAcademicYear";
import { Calendar } from "lucide-react";

interface AcademicYearSelectorProps {
  className?: string;
  showActiveBadge?: boolean;
  onAcademicYearChange?: (academicYearId: string) => void;
}

export function AcademicYearSelector({ 
  className, 
  showActiveBadge = true,
  onAcademicYearChange 
}: AcademicYearSelectorProps) {
  const { 
    academicYears, 
    currentAcademicYear, 
    currentAcademicYearId, 
    changeAcademicYear 
  } = useAcademicYear();

  const handleAcademicYearChange = async (academicYearId: string) => {
    try {
      await changeAcademicYear(academicYearId);
      onAcademicYearChange?.(academicYearId);
    } catch (error) {
      console.error('Failed to change academic year:', error);
    }
  };

  // Function to format academic year name (e.g., "2025/26" -> "AY-25/26")
  const formatAcademicYearName = (yearName: string) => {
    // Extract the last 2 digits from each year part
    const parts = yearName.split('/');
    if (parts.length === 2) {
      const startYear = parts[0].slice(-2);
      const endYear = parts[1].slice(-2);
      return `AY-${startYear}/${endYear}`;
    }
    // Fallback to original name if format is unexpected
    return yearName;
  };

  if (!academicYears || academicYears.length === 0) {
    return (
      <div className={`flex items-center gap-2 text-muted-foreground ${className}`}>
        <Calendar className="h-4 w-4" />
        <span className="text-sm">No academic years configured</span>
      </div>
    );
  }

  return (
    <div className={`flex items-center gap-2 ${className}`}>
      <Select
        value={currentAcademicYearId || ""}
        onValueChange={handleAcademicYearChange}
      >
        <SelectTrigger className="w-auto min-w-[120px] flex items-center gap-2">
          <Calendar className="h-4 w-4 text-muted-foreground" />
          <span className="font-medium">
            {currentAcademicYear ? formatAcademicYearName(currentAcademicYear.name) : "Select AY"}
          </span>
        </SelectTrigger>
        <SelectContent>
          {academicYears.map((year: any) => (
            <SelectItem key={year._id} value={year._id}>
              <div className="flex items-center gap-2">
                <Calendar className="h-4 w-4 text-muted-foreground" />
                <span>{formatAcademicYearName(year.name)}</span>
                {year.isActive && showActiveBadge && (
                  <Badge variant="default" className="text-xs">Active</Badge>
                )}
                {year.isStaging && (
                  <Badge variant="secondary" className="text-xs">Staging</Badge>
                )}
              </div>
            </SelectItem>
          ))}
        </SelectContent>
      </Select>
    </div>
  );
} 