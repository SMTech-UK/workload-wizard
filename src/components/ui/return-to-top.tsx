"use client";

import { Button } from "@/components/ui/button";
import { ChevronUp, Home } from "lucide-react";
import { useState, useEffect } from "react";
import Link from "next/link";

export function ReturnToTop() {
  const [isVisible, setIsVisible] = useState(true); // Temporarily always visible for testing

  useEffect(() => {
    const toggleVisibility = () => {
      console.log("Scroll position:", window.pageYOffset);
      if (window.pageYOffset > 300) {
        console.log("Setting visible to true");
        setIsVisible(true);
      } else {
        console.log("Setting visible to false");
        setIsVisible(false);
      }
    };

    // Call once on mount to check initial state
    toggleVisibility();
    
    window.addEventListener("scroll", toggleVisibility);
    return () => window.removeEventListener("scroll", toggleVisibility);
  }, []);

  const scrollToTop = () => {
    console.log("Scrolling to top");
    window.scrollTo({
      top: 0,
      behavior: "smooth",
    });
  };

  console.log("ReturnToTop component render, isVisible:", isVisible);

  return (
    <>
      {/* Return to Home button - top left */}
      <Link href="/">
        <Button
          size="sm"
          variant="outline"
          className="fixed top-4 left-4 z-50 shadow-lg hover:shadow-xl transition-all duration-200"
          aria-label="Return to home"
        >
          <Home className="h-4 w-4 mr-2" />
          Home
        </Button>
      </Link>

      {/* Return to Top button - bottom left */}
      <Button
        onClick={scrollToTop}
        size="sm"
        className="fixed bottom-4 left-4 z-50 rounded-full shadow-lg hover:shadow-xl transition-all duration-200 bg-blue-500 text-white"
        aria-label="Return to top"
      >
        <ChevronUp className="h-4 w-4" />
      </Button>
    </>
  );
} 