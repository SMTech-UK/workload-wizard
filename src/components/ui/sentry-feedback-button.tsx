"use client";

import { useEffect, useState } from "react";
import * as Sentry from "@sentry/nextjs";
import { Button } from "@/components/ui/button";
import { Bug, MessageCircle } from "lucide-react";
import { useModalAwareButton } from "@/hooks/useModalAwareButton";
import { createPortal } from "react-dom";

export function SentryFeedbackButton() {
  const [feedback, setFeedback] = useState<any>(null);
  const [widget, setWidget] = useState<any>(null);
  const [mounted, setMounted] = useState(false);
  const buttonRef = useModalAwareButton();

  useEffect(() => {
    setMounted(true);
    // Get the feedback instance on the client side
    setFeedback(Sentry.getFeedback());
  }, []);

  const handleFeedbackClick = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
    
    if (!feedback) return;

    if (widget) {
      // If widget is open, close it
      widget.removeFromDom();
      setWidget(null);
    } else {
      // Create and show the form immediately
      feedback.createForm().then((newWidget: any) => {
        newWidget.appendToDom();
        newWidget.open();
        setWidget(newWidget);
      }).catch((error: any) => {
        console.error('Failed to create Sentry feedback form:', error);
      });
    }
  };

  if (!feedback || !mounted) return null;

  const button = (
    <Button
      ref={buttonRef}
      variant="destructive"
      size="lg"
      onClick={handleFeedbackClick}
      onMouseDown={(e) => {
        e.preventDefault();
        e.stopPropagation();
      }}
      className="fixed bottom-4 right-4 z-[99999] shadow-lg"
      style={{
        zIndex: 99999,
        pointerEvents: 'auto',
      }}
      aria-label="Report a bug or give feedback"
    >
      <Bug className="h-4 w-4 mr-2" />
      Report an Issue
    </Button>
  );

  // Render the button in a portal to ensure it's outside modal context
  return createPortal(button, document.body);
} 