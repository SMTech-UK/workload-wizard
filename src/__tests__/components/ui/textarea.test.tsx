import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Textarea } from '@/components/ui/textarea';

// Custom query to find elements by data-slot attribute
const getByDataSlot = (container: HTMLElement, slotName: string) => {
  return container.querySelector(`[data-slot="${slotName}"]`);
};

describe('Textarea Component - Comprehensive UI Testing', () => {
  describe('Basic Rendering and Functionality', () => {
    it('should render textarea with default styling', () => {
      // Arrange
      const { container } = render(<Textarea />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea).toBeInTheDocument();
      expect(textarea).toHaveClass(
        'border-input', 'flex', 'min-h-16', 'w-full', 'rounded-md', 'border', 'bg-transparent', 'px-3', 'py-2', 'text-base'
      );
    });

    it('should render textarea with custom className', () => {
      // Arrange
      const { container } = render(<Textarea className="custom-textarea" />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea).toHaveClass('custom-textarea');
      expect(textarea).toHaveClass('border-input', 'flex', 'min-h-16', 'w-full', 'rounded-md', 'border');
    });

    it('should render textarea with placeholder', () => {
      // Arrange
      const { container } = render(<Textarea placeholder="Enter your message" />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(textarea).toHaveAttribute('placeholder', 'Enter your message');
      expect(textarea).toHaveClass('placeholder:text-muted-foreground');
    });

    it('should render textarea with value', () => {
      // Arrange
      const { container } = render(<Textarea defaultValue="Initial text" />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(textarea).toHaveValue('Initial text');
    });

    it('should render textarea with rows and cols', () => {
      // Arrange
      const { container } = render(<Textarea rows={5} cols={50} />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(textarea).toHaveAttribute('rows', '5');
      expect(textarea).toHaveAttribute('cols', '50');
    });

    it('should render textarea with maxLength', () => {
      // Arrange
      const { container } = render(<Textarea maxLength={100} />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(textarea).toHaveAttribute('maxLength', '100');
    });
  });

  describe('Accessibility Testing', () => {
    it('should have proper semantic structure', () => {
      // Arrange
      const { container } = render(<Textarea />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea?.tagName).toBe('TEXTAREA');
      expect(textarea).toHaveAttribute('data-slot', 'textarea');
    });

    it('should support ARIA attributes', () => {
      // Arrange
      const { container } = render(<Textarea aria-label="Message input" />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea).toHaveAttribute('aria-label', 'Message input');
    });

    it('should support aria-describedby attribute', () => {
      // Arrange
      const { container } = render(<Textarea aria-describedby="textarea-help" />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea).toHaveAttribute('aria-describedby', 'textarea-help');
    });

    it('should support aria-invalid attribute', () => {
      // Arrange
      const { container } = render(<Textarea aria-invalid="true" />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea).toHaveAttribute('aria-invalid', 'true');
      expect(textarea).toHaveClass('aria-invalid:ring-destructive/20', 'aria-invalid:border-destructive');
    });

    it('should support aria-required attribute', () => {
      // Arrange
      const { container } = render(<Textarea aria-required="true" />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea).toHaveAttribute('aria-required', 'true');
    });

    it('should support focus-visible styling', () => {
      // Arrange
      const { container } = render(<Textarea />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea).toHaveClass('focus-visible:border-ring', 'focus-visible:ring-ring/50', 'focus-visible:ring-[3px]');
    });

    it('should support disabled state', () => {
      // Arrange
      const { container } = render(<Textarea disabled />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(textarea).toBeDisabled();
      expect(textarea).toHaveClass('disabled:cursor-not-allowed', 'disabled:opacity-50');
    });
  });

  describe('User Interaction Testing', () => {
    it('should handle text input', async () => {
      // Arrange
      const user = userEvent.setup();
      const { container } = render(<Textarea />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.type(textarea, 'Hello, world!');

      // Assert
      expect(textarea).toHaveValue('Hello, world!');
    });

    it('should handle onChange event', async () => {
      // Arrange
      const user = userEvent.setup();
      const handleChange = jest.fn();
      const { container } = render(<Textarea onChange={handleChange} />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.type(textarea, 'Test');

      // Assert
      expect(handleChange).toHaveBeenCalled();
    });

    it('should handle onFocus event', async () => {
      // Arrange
      const user = userEvent.setup();
      const handleFocus = jest.fn();
      const { container } = render(<Textarea onFocus={handleFocus} />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.click(textarea);

      // Assert
      expect(handleFocus).toHaveBeenCalled();
    });

    it('should handle onBlur event', async () => {
      // Arrange
      const user = userEvent.setup();
      const handleBlur = jest.fn();
      const { container } = render(<Textarea onBlur={handleBlur} />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.click(textarea);
      await user.tab();

      // Assert
      expect(handleBlur).toHaveBeenCalled();
    });

    it('should handle keyboard navigation', async () => {
      // Arrange
      const user = userEvent.setup();
      const { container } = render(<Textarea />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.type(textarea, 'Line 1{Enter}Line 2{Enter}Line 3');

      // Assert
      expect(textarea).toHaveValue('Line 1\nLine 2\nLine 3');
    });

    it('should handle paste events', async () => {
      // Arrange
      const user = userEvent.setup();
      const { container } = render(<Textarea />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.click(textarea);
      await user.paste('Pasted content');

      // Assert
      expect(textarea).toHaveValue('Pasted content');
    });

    it('should not allow input when disabled', async () => {
      // Arrange
      const user = userEvent.setup();
      const { container } = render(<Textarea disabled />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.type(textarea, 'This should not work');

      // Assert
      expect(textarea).toHaveValue('');
      expect(textarea).toBeDisabled();
    });
  });

  describe('Responsive Behavior', () => {
    it('should maintain proper styling on mobile screens', () => {
      // Arrange
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 375,
      });

      const { container } = render(<Textarea />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea).toHaveClass('w-full', 'min-h-16', 'text-base');
    });

    it('should maintain proper styling on tablet screens', () => {
      // Arrange
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 768,
      });

      const { container } = render(<Textarea />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea).toHaveClass('w-full', 'min-h-16', 'md:text-sm');
    });

    it('should maintain proper styling on desktop screens', () => {
      // Arrange
      Object.defineProperty(window, 'innerWidth', {
        writable: true,
        configurable: true,
        value: 1024,
      });

      const { container } = render(<Textarea />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea).toHaveClass('w-full', 'min-h-16', 'md:text-sm');
    });

    it('should handle custom responsive classes', () => {
      // Arrange
      const { container } = render(<Textarea className="md:min-h-20 lg:min-h-24" />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea).toHaveClass('md:min-h-20', 'lg:min-h-24');
    });
  });

  describe('Content and Validation', () => {
    it('should handle empty textarea', () => {
      // Arrange
      const { container } = render(<Textarea />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(textarea).toHaveValue('');
    });

    it('should handle very long text', async () => {
      // Arrange
      const user = userEvent.setup();
      const longText = 'A'.repeat(1000);
      const { container } = render(<Textarea />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.type(textarea, longText);

      // Assert
      expect(textarea).toHaveValue(longText);
    });

    it('should handle multiline text', async () => {
      // Arrange
      const user = userEvent.setup();
      const multilineText = 'Line 1\nLine 2\nLine 3\nLine 4';
      const { container } = render(<Textarea />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.type(textarea, multilineText);

      // Assert
      expect(textarea).toHaveValue(multilineText);
    });

    it('should handle special characters', async () => {
      // Arrange
      const user = userEvent.setup();
      const specialChars = 'Special chars: <>&"\'<>{}[]()@#$%^&*()';
      const { container } = render(<Textarea />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.type(textarea, specialChars);

      // Assert
      expect(textarea).toHaveValue(specialChars);
    });

    it('should handle emoji characters', async () => {
      // Arrange
      const user = userEvent.setup();
      const emojiText = 'Hello 👋 World 🌍 with emojis 😊';
      const { container } = render(<Textarea />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.type(textarea, emojiText);

      // Assert
      expect(textarea).toHaveValue(emojiText);
    });

    it('should respect maxLength constraint', async () => {
      // Arrange
      const user = userEvent.setup();
      const { container } = render(<Textarea maxLength={10} />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.type(textarea, 'This is too long');

      // Assert
      expect(textarea).toHaveValue('This is to');
      expect(textarea).toHaveAttribute('maxLength', '10');
    });

    it('should handle required validation', () => {
      // Arrange
      const { container } = render(<Textarea required />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(textarea).toBeRequired();
    });

    it('should handle readOnly state', () => {
      // Arrange
      const { container } = render(<Textarea readOnly defaultValue="Read only text" />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(textarea).toHaveAttribute('readOnly');
      expect(textarea).toHaveValue('Read only text');
    });
  });

  describe('Edge Cases and Error Handling', () => {
    it('should handle controlled component properly', async () => {
      // Arrange
      const user = userEvent.setup();
      const handleChange = jest.fn();
      const { container, rerender } = render(<Textarea value="" onChange={handleChange} />);

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.type(textarea, 'New text');

      // Assert
      expect(handleChange).toHaveBeenCalled();
      expect(textarea).toHaveValue(''); // Should remain empty as it's controlled

      // Update value
      rerender(<Textarea value="Updated text" onChange={handleChange} />);
      expect(textarea).toHaveValue('Updated text');
    });

    it('should handle multiple textareas', async () => {
      // Arrange
      const user = userEvent.setup();
      const { container } = render(
        <div>
          <Textarea data-testid="textarea1" />
          <Textarea data-testid="textarea2" />
          <Textarea data-testid="textarea3" />
        </div>
      );

      // Act
      const textareas = container.querySelectorAll('[data-slot="textarea"]');
      await user.type(textareas[0] as HTMLTextAreaElement, 'Text 1');
      await user.type(textareas[1] as HTMLTextAreaElement, 'Text 2');
      await user.type(textareas[2] as HTMLTextAreaElement, 'Text 3');

      // Assert
      expect(textareas).toHaveLength(3);
      expect(textareas[0]).toHaveValue('Text 1');
      expect(textareas[1]).toHaveValue('Text 2');
      expect(textareas[2]).toHaveValue('Text 3');
    });

    it('should handle textarea with form context', () => {
      // Arrange
      const { container } = render(
        <form>
          <Textarea name="message" id="message-textarea" />
        </form>
      );

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(textarea).toHaveAttribute('name', 'message');
      expect(textarea).toHaveAttribute('id', 'message-textarea');
    });

    it('should handle textarea with auto-resize behavior', () => {
      // Arrange
      const { container } = render(<Textarea style={{ resize: 'vertical' }} />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(textarea).toHaveStyle({ resize: 'vertical' });
    });

    it('should handle textarea with spellcheck', () => {
      // Arrange
      const { container } = render(<Textarea spellCheck={false} />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(textarea).toHaveAttribute('spellCheck', 'false');
    });
  });

  describe('Snapshot Testing for UI Consistency', () => {
    it('should match snapshot for default textarea', () => {
      // Arrange
      const { container } = render(<Textarea />);

      // Act & Assert
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for textarea with placeholder', () => {
      // Arrange
      const { container } = render(<Textarea placeholder="Enter your message" />);

      // Act & Assert
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for disabled textarea', () => {
      // Arrange
      const { container } = render(<Textarea disabled />);

      // Act & Assert
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for invalid textarea', () => {
      // Arrange
      const { container } = render(<Textarea aria-invalid="true" />);

      // Act & Assert
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for multiple textareas', () => {
      // Arrange
      const { container } = render(
        <div>
          <Textarea placeholder="Textarea 1" />
          <Textarea placeholder="Textarea 2" />
          <Textarea placeholder="Textarea 3" />
        </div>
      );

      // Act & Assert
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('Performance Testing', () => {
    it('should handle rapid typing efficiently', async () => {
      // Arrange
      const user = userEvent.setup();
      const { container } = render(<Textarea />);

      const startTime = performance.now();

      // Act
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      await user.type(textarea, 'A'.repeat(100));

      const endTime = performance.now();

      // Assert
      expect(textarea).toHaveValue('A'.repeat(100));
      expect(endTime - startTime).toBeLessThan(1000); // Should type within 1 second
    });

    it('should handle multiple textarea renders efficiently', () => {
      // Arrange
      const textareas = Array.from({ length: 100 }, (_, i) => `textarea-${i + 1}`);

      const startTime = performance.now();

      // Act
      render(
        <div>
          {textareas.map((id, index) => (
            <Textarea key={index} data-testid={id} placeholder={`Placeholder ${index + 1}`} />
          ))}
        </div>
      );

      const endTime = performance.now();

      // Assert
      expect(screen.getByTestId('textarea-1')).toBeInTheDocument();
      expect(screen.getByTestId('textarea-100')).toBeInTheDocument();
      expect(endTime - startTime).toBeLessThan(1000); // Should render within 1 second
    });

    it('should handle className changes efficiently', () => {
      // Arrange
      const { rerender, container } = render(<Textarea className="initial-class" />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea');
      expect(textarea).toHaveClass('initial-class');

      // Change className
      rerender(<Textarea className="updated-class" />);

      const updatedTextarea = getByDataSlot(container, 'textarea');
      expect(updatedTextarea).toHaveClass('updated-class');
      expect(updatedTextarea).not.toHaveClass('initial-class');
    });

    it('should handle value changes efficiently', () => {
      // Arrange
      const { rerender, container } = render(<Textarea value="Initial value" />);

      // Act & Assert
      const textarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(textarea).toHaveValue('Initial value');

      // Change value
      rerender(<Textarea value="Updated value" />);

      const updatedTextarea = getByDataSlot(container, 'textarea') as HTMLTextAreaElement;
      expect(updatedTextarea).toHaveValue('Updated value');
    });
  });
}); 