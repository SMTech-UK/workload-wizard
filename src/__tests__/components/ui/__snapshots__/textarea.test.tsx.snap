// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for default textarea 1`] = `
<textarea
  class="border-input placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 flex field-sizing-content min-h-16 w-full rounded-md border bg-transparent px-3 py-2 text-base shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
  data-slot="textarea"
/>
`;

exports[`Textarea Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for disabled textarea 1`] = `
<textarea
  class="border-input placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 flex field-sizing-content min-h-16 w-full rounded-md border bg-transparent px-3 py-2 text-base shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
  data-slot="textarea"
  disabled=""
/>
`;

exports[`Textarea Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for invalid textarea 1`] = `
<textarea
  aria-invalid="true"
  class="border-input placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 flex field-sizing-content min-h-16 w-full rounded-md border bg-transparent px-3 py-2 text-base shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
  data-slot="textarea"
/>
`;

exports[`Textarea Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for multiple textareas 1`] = `
<div>
  <textarea
    class="border-input placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 flex field-sizing-content min-h-16 w-full rounded-md border bg-transparent px-3 py-2 text-base shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
    data-slot="textarea"
    placeholder="Textarea 1"
  />
  <textarea
    class="border-input placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 flex field-sizing-content min-h-16 w-full rounded-md border bg-transparent px-3 py-2 text-base shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
    data-slot="textarea"
    placeholder="Textarea 2"
  />
  <textarea
    class="border-input placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 flex field-sizing-content min-h-16 w-full rounded-md border bg-transparent px-3 py-2 text-base shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
    data-slot="textarea"
    placeholder="Textarea 3"
  />
</div>
`;

exports[`Textarea Component - Comprehensive UI Testing Snapshot Testing for UI Consistency should match snapshot for textarea with placeholder 1`] = `
<textarea
  class="border-input placeholder:text-muted-foreground focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive dark:bg-input/30 flex field-sizing-content min-h-16 w-full rounded-md border bg-transparent px-3 py-2 text-base shadow-xs transition-[color,box-shadow] outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 md:text-sm"
  data-slot="textarea"
  placeholder="Enter your message"
/>
`;
